### Cardiac Arrhythmia Multy-Class Classification 

Analyze data and address missing data if there is any. 

Decide aboute a good evaluation strategy and justify your choice. 

Find the best parameters for the following classification models: 
- KNN classifcation x
- Logistic Regressionx
- Linear Supprt Vector Machine x
- Kerenilzed Support Vector Machine x
- Decision Tree x
- Random Forest x

Then use different bagging and boosting methods to boost the results? Do you see any significant change? Why or why not? 

Next, use data reduction method you have learned in class to reduce the size of data, and agian try above models. Do you get better results? Justify your answer. 

<font color = 'red'>Due date for full credit: April 4, 11:59 PM
    <br>
    Due date for partial credit: April 6, 11:59 PM.
    <br> No submission will be accepted after April 6. 
    <br> Please note that your term paper is also due April 6. 
    </font>

### Data Import, and Data Preparation

In [4]:
import pandas as pd

import numpy as np
import os
os.chdir('/Users/dhruvjoshi/Desktop/ML')

rawdata = pd.read_csv("cardiac_arrhythmia.csv",header=None)


In [5]:
rawdata

0    1    2    3    4    5    6    7    8    9   ...   270   271   272  \
0     75    0  190   80   91  193  371  174  121  -16 ...   0.0   9.0  -0.9   
1     56    1  165   64   81  174  401  149   39   25 ...   0.0   8.5   0.0   
2     54    0  172   95  138  163  386  185  102   96 ...   0.0   9.5  -2.4   
3     55    0  175   94  100  202  380  179  143   28 ...   0.0  12.2  -2.2   
4     75    0  190   80   88  181  360  177  103  -16 ...   0.0  13.1  -3.6   
5     13    0  169   51  100  167  321  174   91  107 ...  -0.6  12.2  -2.8   
6     40    1  160   52   77  129  377  133   77   77 ...   0.0   6.5   0.0   
7     49    1  162   54   78    0  376  157   70   67 ...   0.0   8.2  -1.9   
8     44    0  168   56   84  118  354  160   63   61 ...   0.0   7.0  -1.3   
9     50    1  167   67   89  130  383  156   73   85 ...  -0.6  10.8  -1.7   
10    62    0  170   72  102  135  401  156   83   72 ...  -0.5   9.0  -2.0   
11    45    1  165   86   77  143  373  150   65   12 ...   0.0   4.4  -2.2   
12    54    1  172   58   78  155  382  163   81  -24 ...   0.0   6.3  -2.1   
13    30    0  170   73   91  180  355  157  104   68 ...  -0.9  12.3   0.0   
14    44    1  160   88   77  158  399  163   94   46 ...  -0.6  12.4   0.0   
15    47    1  150   48   75  132  350  169   65   36 ...   0.0   7.7  -0.8   
16    47    0  171   59   82  145  347  169   61   77 ...   0.0   9.4  -1.7   
17    46    1  158   58   70  120  353  122   52   57 ...   0.0   6.6   0.0   
18    73    0  165   63   91  154  392  175   83   73 ...   0.0   5.7   0.0   
19    57    1  166   72   82  181  399  158   79  -12 ...   0.0   7.7  -0.9   
20    28    1  160   58   83  251  383  189  183   50 ...  -0.6   9.1  -1.4   
21    45    0  169   67   90  122  336  177   78   81 ...  -0.6   8.3  -1.8   
22    36    1  153   75   71  132  364  169   82   62 ...   0.0   8.9  -1.0   
23    57    1  165   59   75  157  406  143   92    4 ...   0.0   6.7  -0.5   
24    40    1  153   55   82  140  388  149   82   52 ...   0.0  13.6   0.0   
25    44    0  169   80  109  128  382  195   60  -34 ...   0.0   6.9   0.0   
26    34    0  170   73   94  186  373  224  125   90 ...   0.0  15.3  -1.1   
27    31    1  160   54   95  161  407  168   83   10 ...   0.0  12.7  -1.8   
28    56    1  164   65   90  164  420  381   99   -8 ...   0.0   5.4   0.0   
29    51    1  160   83   96  147  400  301   82  -37 ...   0.0   7.3  -3.9   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ... ...   ...   ...   ...   
422   29    1  162   57   83  164  359  154   69   64 ...   0.0  14.1  -2.2   
423   51    0  186   95   94  203  367  171  106   -7 ...   0.0   9.6  -3.5   
424    7    0  119   21  140  157  438  226   81  -40 ...   0.0  10.0  -2.1   
425   36    0  171   93   87  150  362  177   96   44 ...   0.0  10.3  -0.8   
426   35    1  160   53   55  163  340  162  102   40 ...   0.0   8.7  -0.5   
427   58    0  160   65  133  148  417  260   92 -158 ...  -0.4   6.4  -3.5   
428   64    0  160   63   83    0  364  120   90   29 ...   0.0   6.7  -0.4   
429    8    1  130   24   77  125  358  159   70   87 ...   0.0  11.3  -2.1   
430   11    0  138   29  123  145  361  221   80  112 ...  -3.4  19.6  -4.2   
431   47    0  166   56   79  145  381  173  101   52 ...   0.0   8.5   0.0   
432   11    0  140   42   88  123  362  228   81  -18 ...   0.0  17.1  -7.1   
433   70    0  167   60   80  149  290  128   93  -67 ...   0.0   2.7  -5.4   
434   20    0  178   65   88  155  360  163   71  -22 ...  -0.5  10.2   0.0   
435   39    1  164   62   79  155  367  153   95   50 ...   0.0   9.7  -0.7   
436   32    1  164   57   77  144  340  148   82   27 ...  -0.6   9.9  -0.6   
437   35    1  155   63   87  142  391  137   88   66 ...   0.0  10.7   0.0   
438   37    0  175   82   88  146  357  179   72    1 ...  -0.4  13.5  -1.2   
439   49    1  168   66   94  170  383  152  115   92 ...   0.0   8.2  -0.7   
440   37    0  176   72   88  153  389  172   89   67 .

### Looking in the spreadsheet, we saw some missing values, let's take a look

In [6]:
rawdata[13].head()

0     ?
1     ?
2    23
3     ?
4     ?
Name: 13, dtype: object

In [7]:
rawdata.describe()

0           1           2           3           4           5    \
count  452.000000  452.000000  452.000000  452.000000  452.000000  452.000000   
mean    46.471239    0.550885  166.188053   68.170354   88.920354  155.152655   
std     16.466631    0.497955   37.170340   16.590803   15.364394   44.842283   
min      0.000000    0.000000  105.000000    6.000000   55.000000    0.000000   
25%     36.000000    0.000000  160.000000   59.000000   80.000000  142.000000   
50%     47.000000    1.000000  164.000000   68.000000   86.000000  157.000000   
75%     58.000000    1.000000  170.000000   79.000000   94.000000  175.000000   
max     83.000000    1.000000  780.000000  176.000000  188.000000  524.000000   

              6           7           8           9       ...             270  \
count  452.000000  452.000000  452.000000  452.000000     ...      452.000000   
mean   367.207965  169.949115   90.004425   33.676991     ...       -0.278982   
std     33.385421   35.633072   25.826643   45.431434     ...        0.548876   
min    232.000000  108.000000    0.000000 -172.000000     ...       -4.100000   
25%    350.000000  148.000000   79.000000    3.750000     ...       -0.425000   
50%    367.000000  162.000000   91.000000   40.000000     ...        0.000000   
75%    384.000000  179.000000  102.000000   66.000000     ...        0.000000   
max    509.000000  381.000000  205.000000  169.000000     ...        0.000000   

              271         272         273    274         275         276  \
count  452.000000  452.000000  452.000000  452.0  452.000000  452.000000   
mean     9.048009   -1.457301    0.003982    0.0    0.514823    1.222345   
std      3.472862    2.002430    0.050118    0.0    0.347531    1.426052   
min      0.000000  -28.600000    0.000000    0.0   -0.800000   -6.000000   
25%      6.600000   -2.100000    0.000000    0.0    0.400000    0.500000   
50%      8.800000   -1.100000    0.000000    0.0    0.500000    1.350000   
75%     11.200000    0.000000    0.000000    0.0    0.700000    2.100000   
max     23.600000    0.000000    0.800000    0.0    2.400000    6.000000   

              277         278         279  
count  452.000000  452.000000  452.000000  
mean    19.326106   29.473230    3.880531  
std     13.503922   18.493927    4.407097  
min    -44.200000  -38.600000    1.000000  
25%     11.450000   17.550000    1.000000  
50%     18.100000   27.900000    1.000000  
75%     25.825000   41.125000    6.000000  
max     88.800000  115.900000   16.000000  

[8 rows x 275 columns]

In [8]:
description=rawdata.describe()

In [9]:
countdata = description.iloc[[0]].values

### Note, using isnull() does not show the missing values, since they are not missing, but rather populated with '?' strings

In [10]:
rawdata[13].isnull().head()

0    False
1    False
2    False
3    False
4    False
Name: 13, dtype: bool

### Looking at the info() we see 5 object columns, these must be where the '?' values are found. Looking at sample data from those columns, they are also intended to be numeric, so we convert the dataframe to numeric in order to reveal the presence of NaNs

In [11]:
rawdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Columns: 280 entries, 0 to 279
dtypes: float64(116), int64(159), object(5)
memory usage: 988.8+ KB


In [12]:
convertdata = rawdata.replace({'?': np.nan}).astype(float)

### These are the column indices with missing values present, as well as how many observations they actually contain

In [13]:
convertdata.columns[convertdata.isnull().any()].tolist()

[10, 11, 12, 13, 14]

In [14]:
description = convertdata.describe()
countdata = description.iloc[[0]].values
(countdata[countdata<452])

array([444., 430., 451.,  76., 451.])

### Most of the columns are missing less than 10% of the overall dataset. However, column 13 (the QRST value) is missing a very significant portion of the data. Because of this, our recommendation is to completely remove the QRST variable from the dataset, and to only trim off the observations that have missing values within the other columns (10,11,12, and 14, which are QRS, T, P, and J, respectively). Our suggestion of dropping column 13 and the remaining observations with NaN stem from the fact that we do not possess the proper medical background to fully understand the variables that we are dealing with, and therefore, do not have a good understanding of the impact of imputations that use more assumptions.

In [15]:
del convertdata[13]

In [16]:
cleandata = convertdata.dropna()

In [17]:
cleandata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 420 entries, 0 to 451
Columns: 279 entries, 0 to 279
dtypes: float64(279)
memory usage: 918.8 KB


### After doing the missing data drop, we still have 420 observations, this means we only dropped about 6% of the data, which is still well within the safe margins recommended. 

In [18]:
cleandata.describe()

0           1           2           3           4           5    \
count  420.000000  420.000000  420.000000  420.000000  420.000000  420.000000   
mean    46.050000    0.559524  166.159524   67.657143   88.900000  158.935714   
std     16.387168    0.497036   38.496515   16.633967   15.651774   36.612571   
min      0.000000    0.000000  105.000000    6.000000   55.000000    0.000000   
25%     36.000000    0.000000  160.000000   59.000000   80.000000  143.000000   
50%     46.000000    1.000000  164.000000   68.000000   86.000000  157.000000   
75%     58.000000    1.000000  170.000000   78.000000   94.000000  174.000000   
max     83.000000    1.000000  780.000000  176.000000  188.000000  524.000000   

              6           7           8           9       ...             270  \
count  420.000000  420.000000  420.000000  420.000000     ...      420.000000   
mean   367.559524  170.166667   91.876190   33.273810     ...       -0.274762   
std     33.129078   35.837885   21.822541   46.094422     ...        0.530125   
min    232.000000  111.000000    0.000000 -172.000000     ...       -4.100000   
25%    350.000000  148.000000   79.750000    2.750000     ...       -0.425000   
50%    367.500000  162.500000   91.000000   40.000000     ...        0.000000   
75%    384.000000  178.000000  102.000000   66.000000     ...        0.000000   
max    509.000000  381.000000  205.000000  169.000000     ...        0.000000   

              271         272         273    274         275         276  \
count  420.000000  420.000000  420.000000  420.0  420.000000  420.000000   
mean     9.022619   -1.487143    0.003095    0.0    0.544286    1.270952   
std      3.452782    2.046704    0.045984    0.0    0.323009    1.410061   
min      0.000000  -28.600000    0.000000    0.0   -0.800000   -6.000000   
25%      6.600000   -2.100000    0.000000    0.0    0.400000    0.600000   
50%      8.700000   -1.200000    0.000000    0.0    0.500000    1.400000   
75%     11.200000    0.000000    0.000000    0.0    0.700000    2.100000   
max     23.600000    0.000000    0.800000    0.0    2.400000    6.000000   

              277         278         279  
count  420.000000  420.000000  420.000000  
mean    19.109524   29.618333    3.685714  
std     13.651475   18.584817    4.189005  
min    -44.200000  -38.600000    1.000000  
25%     11.100000   17.750000    1.000000  
50%     18.000000   28.100000    1.000000  
75%     25.800000   41.225000    6.000000  
max     88.800000  115.900000   16.000000  

[8 rows x 279 columns]

### Next, for our evaluation strategy, we will make a new variable to represent the target variable, in which we solely have whether or not a patient's heart is normal, or exhibiting arrythmia. 

### This would be used for models that are best used for binary classification, to help determine whether the attention of a doctor is necessary. In addition, this would be the simplest model - if this model proves useful, pursuing a more complex model would be permissible. 

### On the otherhand, if we tried to first pursue a more complex model, even if the model fails, we would need to backtrack and pursue a simpler model of just diagnosing the presence of arrhythmia. Furthermore, trying to specify any further can lead to a class imbalance issue.

In [19]:
cleandata[280] = cleandata[279]!=1

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [20]:
cleandata[280].head()

0     True
1     True
2     True
3    False
5     True
Name: 280, dtype: bool

In [21]:
cleandata[280].describe()

count       420
unique        2
top       False
freq        237
Name: 280, dtype: object

### So we can see that there are 183 occurrences of arrythmia, and 237 occurrences of regular heart rhythms. Therefore, we determine that, for now, there is no reason to believe that there are any class imbalance problems for the binary classification model - however, it will be something we might need to be wary of for the multi-classification models.

### Lastly, let's separate the data so we have X, and y:

In [22]:
Xdata = cleandata.iloc[:,0:278]

In [23]:
ydatamulti = cleandata.iloc[:,278]

In [24]:
ydatabinary = cleandata.iloc[:,279]

## Let us now build some models

### First, let's build a simple KNN model:

### First, we'd want to split and more importantly SCALE the data, so that no variable has any extra weight than any other variable based on their own scales

In [25]:
from sklearn.model_selection import train_test_split
Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)


from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
Xtrainscaled = scaler.fit_transform(Xtrain)
Xtestscaled = scaler.transform(Xtest)


In [26]:
from sklearn.neighbors import KNeighborsClassifier

for k in range(1,20):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(Xtrainscaled,ytrain)
    print(k)
    print("Training Accuracy:")
    print(knn.score(Xtrainscaled,ytrain))
    print("Test Accuracy:")
    print(knn.score(Xtestscaled,ytest))


1
Training Accuracy:
1.0
Test Accuracy:
0.5523809523809524
2
Training Accuracy:
0.7206349206349206
Test Accuracy:
0.6666666666666666
3
Training Accuracy:
0.7111111111111111
Test Accuracy:
0.6666666666666666
4
Training Accuracy:
0.6730158730158731
Test Accuracy:
0.6571428571428571
5
Training Accuracy:
0.6571428571428571
Test Accuracy:
0.6476190476190476
6
Training Accuracy:
0.6349206349206349
Test Accuracy:
0.6666666666666666
7
Training Accuracy:
0.6253968253968254
Test Accuracy:
0.6476190476190476
8
Training Accuracy:
0.6158730158730159
Test Accuracy:
0.638095238095238
9
Training Accuracy:
0.6126984126984127
Test Accuracy:
0.6666666666666666
10
Training Accuracy:
0.6
Test Accuracy:
0.6571428571428571
11
Training Accuracy:
0.6063492063492063
Test Accuracy:
0.6571428571428571
12
Training Accuracy:
0.5904761904761905
Test Accuracy:
0.6476190476190476
13
Training Accuracy:
0.5841269841269842
Test Accuracy:
0.6476190476190476
14
Training Accuracy:
0.5777777777777777
Test Accuracy:
0.6380952

### It's clear that testing accuracy performs worse on any k larger than 9. Between k=1,9 we have a few with a testing accuracy of 66%. One could argue the simplest k would be best, but with a k=3, we have a relatively high test accuracy, and a training accuracy that is similar, meaning we are most likely not under or overfitting. This model would best generalize into new datasets. Still, the accuracy is quite low overall, so this model may perform better within an ensemble model.

In [27]:
knnFinal = KNeighborsClassifier(n_neighbors=3)
knnFinal.fit(Xtrainscaled,ytrain)
print('k:',3)
print("Training Accuracy:")
print(knnFinal.score(Xtrainscaled,ytrain))
print("Test Accuracy:")
print(knnFinal.score(Xtestscaled,ytest))

k: 3
Training Accuracy:
0.7111111111111111
Test Accuracy:
0.6666666666666666


### Next, let's try a Decision Tree Model

In [30]:
from sklearn.tree import DecisionTreeClassifier

Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)


for d in range (1,20):
    tree = DecisionTreeClassifier(max_depth = d).fit(Xtrain,ytrain)
    print(d)
    print(tree.score(Xtrain,ytrain))
    print(tree.score(Xtest,ytest))

1
0.6126984126984127
0.6666666666666666
2
0.6476190476190476
0.6761904761904762
3
0.7111111111111111
0.6952380952380952
4
0.7619047619047619
0.7142857142857143
5
0.8158730158730159
0.7238095238095238
6
0.8539682539682539
0.7047619047619048
7
0.8984126984126984
0.7428571428571429
8
0.926984126984127
0.6952380952380952
9
0.9333333333333333
0.6857142857142857
10
0.9523809523809523
0.6952380952380952
11
0.9682539682539683
0.7142857142857143
12
0.9746031746031746
0.6761904761904762
13
0.9777777777777777
0.6761904761904762
14
0.9809523809523809
0.6761904761904762
15
0.9809523809523809
0.6476190476190476
16
0.9841269841269841
0.638095238095238
17
0.9873015873015873
0.6857142857142857
18
0.9904761904761905
0.6285714285714286
19
0.9936507936507937
0.6571428571428571


### When we first ran the decision tree classifier without giving a max decision tree depth limit, the test accuracy was in the low 60s. Looking at the above results, we see that any depth greater than 8 results in overfitting of the model. Looking at the test accuracy, a maximum depth of 7 or 8 would best generalize to new datasets.

In [33]:
treeFinal = DecisionTreeClassifier(max_depth = 7)
treeFinal.fit(Xtrain,ytrain)
print('Max Depth:',7)
print('Training Score:',treeFinal.score(Xtrain,ytrain))
print('Testing Score:',treeFinal.score(Xtest,ytest))

Max Depth: 7
Training Score: 0.8984126984126984
Testing Score: 0.7333333333333333


### Next, let's examine a linear SVM classification model

In [34]:
from sklearn.svm import LinearSVC

Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)

scaler = MinMaxScaler()
Xtrainscaled = scaler.fit_transform(Xtrain)
Xtestscaled = scaler.transform(Xtest)

carray = np.linspace(0.1,1,10)
for c in carray:
    LSVM = LinearSVC(C=c,random_state=67).fit(Xtrainscaled,ytrain)
    print("C",c)
    print("training score",LSVM.score(Xtrainscaled,ytrain))
    print("testing score",LSVM.score(Xtestscaled,ytest))
for c in range(2,10):
    LSVM = LinearSVC(C=c,random_state=67).fit(Xtrainscaled,ytrain)
    print("C",c)
    print("training score",LSVM.score(Xtrainscaled,ytrain))
    print("testing score",LSVM.score(Xtestscaled,ytest))

C 0.1
training score 0.819047619047619
testing score 0.7428571428571429
C 0.2
training score 0.873015873015873
testing score 0.7523809523809524
C 0.30000000000000004
training score 0.8952380952380953
testing score 0.7523809523809524
C 0.4
training score 0.9111111111111111
testing score 0.7428571428571429
C 0.5
training score 0.9206349206349206
testing score 0.7428571428571429
C 0.6
training score 0.926984126984127
testing score 0.7523809523809524
C 0.7000000000000001
training score 0.9396825396825397
testing score 0.7523809523809524
C 0.8
training score 0.9428571428571428
testing score 0.7428571428571429
C 0.9
training score 0.9555555555555556
testing score 0.7333333333333333
C 1.0
training score 0.9587301587301588
testing score 0.7333333333333333
C 2
training score 0.9746031746031746
testing score 0.7142857142857143
C 3
training score 0.9841269841269841
testing score 0.6857142857142857
C 4
training score 0.9904761904761905
testing score 0.6761904761904762
C 5
training score 0.99365079

### The best training and testing scores result from a C around 0.7

In [35]:
LSVMFinal = LinearSVC(C=0.7,random_state=67)
LSVMFinal.fit(Xtrainscaled,ytrain)
print("C",0.7)
print("training score",LSVM.score(Xtrainscaled,ytrain))
print("testing score",LSVM.score(Xtestscaled,ytest))

C 0.7
training score 0.9904761904761905
testing score 0.6666666666666666


### Next, let's try out a kernelized SVM

In [36]:
from sklearn.svm import SVC

Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)

scaler = MinMaxScaler()
Xtrainscaled = scaler.fit_transform(Xtrain)
Xtestscaled = scaler.transform(Xtest)

garray = (0.1,1,10,50,100)
carray = (0.1,1,10,50,100)
for c in carray:
    for g in garray:
        KSVM = SVC(kernel = 'rbf',gamma=g,C=c).fit(Xtrainscaled,ytrain)
        print("C:",c)
        print("gamma",g)
        print("Training",KSVM.score(Xtrainscaled,ytrain))
        print("Test",KSVM.score(Xtestscaled,ytest))


C: 0.1
gamma 0.1
Training 0.5428571428571428
Test 0.6285714285714286
C: 0.1
gamma 1
Training 0.5428571428571428
Test 0.6285714285714286
C: 0.1
gamma 10
Training 0.5428571428571428
Test 0.6285714285714286
C: 0.1
gamma 50
Training 0.5428571428571428
Test 0.6285714285714286
C: 0.1
gamma 100
Training 0.5428571428571428
Test 0.6285714285714286
C: 1
gamma 0.1
Training 0.7555555555555555
Test 0.6857142857142857
C: 1
gamma 1
Training 0.9555555555555556
Test 0.6285714285714286
C: 1
gamma 10
Training 1.0
Test 0.6285714285714286
C: 1
gamma 50
Training 1.0
Test 0.6285714285714286
C: 1
gamma 100
Training 1.0
Test 0.6285714285714286
C: 10
gamma 0.1
Training 0.9746031746031746
Test 0.7238095238095238
C: 10
gamma 1
Training 1.0
Test 0.6190476190476191
C: 10
gamma 10
Training 1.0
Test 0.6285714285714286
C: 10
gamma 50
Training 1.0
Test 0.6285714285714286
C: 10
gamma 100
Training 1.0
Test 0.6285714285714286
C: 50
gamma 0.1
Training 1.0
Test 0.7047619047619048
C: 50
gamma 1
Training 1.0
Test 0.6190476190

### From our findings, we see that the best general parameters to use in order to have a good generalizing model running an SVM with an rbf kernel are a C = 10, and a gamma = 0.1

In [37]:
KSVMFinal = SVC(kernel = 'rbf',gamma=0.1,C=10)
KSVMFinal.fit(Xtrainscaled,ytrain)
print("C:",10)
print("gamma",0.1)
print("Training",KSVMFinal.score(Xtrainscaled,ytrain))
print("Test",KSVMFinal.score(Xtestscaled,ytest))

C: 10
gamma 0.1
Training 0.9746031746031746
Test 0.7238095238095238


### Next let's try making a random forest model

In [38]:
from sklearn.ensemble import RandomForestClassifier

Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)
rforest = RandomForestClassifier().fit(Xtrain,ytrain)

print('Training score:',rforest.score(Xtrain,ytrain))
print('Testing score:',rforest.score(Xtest,ytest))

Training score: 0.9936507936507937
Testing score: 0.7523809523809524


### This looks good, so we'll keep it as it is

In [39]:
rforestFinal = RandomForestClassifier()
rforestFinal.fit(Xtrain,ytrain)
print('Training score:',rforestFinal.score(Xtrain,ytrain))
print('Testing score:',rforestFinal.score(Xtest,ytest))

Training score: 0.9968253968253968
Testing score: 0.7142857142857143


### Logistic Regression

In [40]:
from sklearn.linear_model import LogisticRegression

Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)

for c in carray:
    logit = LogisticRegression(C=c).fit(Xtrain,ytrain)
    print('c parameter:',c)
    print('training score:',logit.score(Xtrain,ytrain))
    print('testing score:',logit.score(Xtest,ytest))

c parameter: 0.1
training score: 0.9904761904761905
testing score: 0.638095238095238
c parameter: 1
training score: 1.0
testing score: 0.6095238095238096
c parameter: 10
training score: 1.0
testing score: 0.580952380952381
c parameter: 50
training score: 1.0
testing score: 0.580952380952381
c parameter: 100
training score: 1.0
testing score: 0.580952380952381


### Looks like we're overfitting past c=1, so we'll try exploring more values in between 0 and 1 (We won't bore you with the iterations, so we'll show you the last one)

In [41]:
carray = (0.0001,0.001,0.01,0.1)
for c in carray:
    logit = LogisticRegression(C=c).fit(Xtrain,ytrain)
    print('c paramter:',c)
    print('training score:',logit.score(Xtrain,ytrain))
    print('testing score:',logit.score(Xtest,ytest))

c paramter: 0.0001
training score: 0.7873015873015873
testing score: 0.7428571428571429
c paramter: 0.001
training score: 0.9015873015873016
testing score: 0.7238095238095238
c paramter: 0.01
training score: 0.9777777777777777
testing score: 0.6857142857142857
c paramter: 0.1
training score: 0.9904761904761905
testing score: 0.638095238095238


### Our best test scores result from a very low C, so we'll recommend that for now. Let's explore changing the multi-class option to multinomial next.

In [42]:
carray = (0.001,0.01,0.1,1)

multiarray = ('newton-cg','sag','saga','lbfgs')

for m in multiarray:
        for c in carray:
            logit = LogisticRegression(C=c,multi_class='multinomial',solver=m,max_iter=10000).fit(Xtrain,ytrain)
            print('solver model=',m)
            print('c paramter:',c)
            print('train score:',logit.score(Xtrain,ytrain))
            print('test score:',logit.score(Xtest,ytest))

solver model= newton-cg
c paramter: 0.001
train score: 0.9396825396825397
test score: 0.7333333333333333
solver model= newton-cg
c paramter: 0.01
train score: 1.0
test score: 0.6761904761904762
solver model= newton-cg
c paramter: 0.1
train score: 1.0
test score: 0.6476190476190476
solver model= newton-cg
c paramter: 1
train score: 1.0
test score: 0.6476190476190476
solver model= sag
c paramter: 0.001
train score: 0.9365079365079365
test score: 0.7523809523809524
solver model= sag
c paramter: 0.01
train score: 0.9968253968253968
test score: 0.6857142857142857
solver model= sag
c paramter: 0.1
train score: 1.0
test score: 0.6571428571428571
solver model= sag
c paramter: 1
train score: 1.0
test score: 0.6666666666666666
solver model= saga
c paramter: 0.001
train score: 0.9365079365079365
test score: 0.7523809523809524
solver model= saga
c paramter: 0.01
train score: 0.9936507936507937
test score: 0.6857142857142857
solver model= saga
c paramter: 0.1
train score: 1.0
test score: 0.67619047

### Our recommendation for a multiclass logit model is to use a C parameter of 0.001, with a solver model of 'sag','saga',or 'lbfgs.' This results in a training score around 93%, and a testing score of 75%.

In [43]:
logitFinal = LogisticRegression(C=0.001,multi_class='multinomial',solver='saga',max_iter=10000)
logitFinal.fit(Xtrain,ytrain)
print('solver model:','saga, sag, or lbfgs')
print('c parameter:',0.001)
print('train score:',logitFinal.score(Xtrain,ytrain))
print('test score:',logitFinal.score(Xtest,ytest))

solver model: saga, sag, or lbfgs
c parameter: 0.001
train score: 0.9365079365079365
test score: 0.7523809523809524


## Next, we will try using our previous models in a voting classifier.


### First we'll try a hard voting classifier using our top three performers.

In [44]:
from sklearn.ensemble import VotingClassifier

Xtrain,Xtest,ytrain,ytest = train_test_split(Xdata,ydatamulti,random_state=0)

train = []
test = []

for i in range(100):
    voteHard=VotingClassifier(estimators=[('logit',logitFinal),("randomForest",rforestFinal),('KSVM',KSVMFinal)],voting='hard')
    voteHard.fit(Xtrain,ytrain)
    train.append(voteHard.score(Xtrain,ytrain))
    test.append(voteHard.score(Xtest,ytest))
    
print('avg training score:',np.mean(train))
print('avg test score:',np.mean(test))

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

avg training score: 0.9946666666666666
avg test score: 0.7427619047619046


/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


### Let's also try it with our lightest models, and see what results

In [ ]:
train = []
test = []

for i in range(100):
    voteHard2=VotingClassifier(estimators=[('logit',logitFinal),("KNN",knnFinal),('tree',treeFinal)],voting='hard')
    voteHard2.fit(Xtrain,ytrain)
    train.append(voteHard2.score(Xtrain,ytrain))
    test.append(voteHard2.score(Xtest,ytest))
    
print('avg training score:',np.mean(train))
print('avg test score:',np.mean(test))

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

avg training score: 0.882539682539683
avg test score: 0.7478095238095237


/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


### Note very interestingly enough that even though the models used for the second voting classifier each perform worse on average than the ones used on the first voting classifier, the second one performs equally as well or even better when aggregated.

### Next, let's try using a soft voting method with models that provide prediction probabilities

In [ ]:
train = []
test = []

KSVMFinalp = SVC(kernel = 'rbf',gamma=0.1,C=10, probability=True)

for i in range(100):
    voteSoft=VotingClassifier(estimators=[('logit',logitFinal),("randomForest",rforestFinal),('KSVM',KSVMFinalp)],voting='soft')
    voteSoft.fit(Xtrain,ytrain)
    train.append(voteSoft.score(Xtrain,ytrain))
    test.append(voteSoft.score(Xtest,ytest))
    
print('avg training score:',np.mean(train))
print('avg test score:',np.mean(test))

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/dhruvjoshi/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of

### Let's also see what happens if we perform bagging methods on our original final models:

In [ ]:
#from sklearn.ensemble import BaggingClassifier

models = [logitFinal,rforestFinal,KSVMFinal,LSVMFinal,treeFinal,knnFinal]

#for model in models:
#bag = BaggingClassifier(logitFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
#bag.fit(Xtrain,ytrain)
#print('model:',model)
#print('training score:',bag.score(Xtrain,ytrain))
#print('testing score:',bag.score(Xtest,ytest))

In [ ]:
from sklearn.ensemble import BaggingClassifier

bag = BaggingClassifier(logitFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
bag.fit(Xtrain,ytrain)
print('model:','logit')
print('training score:',bag.score(Xtrain,ytrain))
print('testing score:',bag.score(Xtest,ytest))

In [ ]:
bag = BaggingClassifier(rforestFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
bag.fit(Xtrain,ytrain)
print('model:','random forest')
print('training score:',bag.score(Xtrain,ytrain))
print('testing score:',bag.score(Xtest,ytest))

In [ ]:
bag = BaggingClassifier(KSVMFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
bag.fit(Xtrain,ytrain)
print('model:','KSVM')
print('training score:',bag.score(Xtrain,ytrain))
print('testing score:',bag.score(Xtest,ytest))

In [ ]:
bag = BaggingClassifier(LSVMFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
bag.fit(Xtrain,ytrain)
print('model:','LSVM')
print('training score:',bag.score(Xtrain,ytrain))
print('testing score:',bag.score(Xtest,ytest))

In [ ]:
bag = BaggingClassifier(treeFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
bag.fit(Xtrain,ytrain)
print('model:','Decision Tree')
print('training score:',bag.score(Xtrain,ytrain))
print('testing score:',bag.score(Xtest,ytest))

In [ ]:
bag = BaggingClassifier(knnFinal,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
bag.fit(Xtrain,ytrain)
print('model:','KNN')
print('training score:',bag.score(Xtrain,ytrain))
print('testing score:',bag.score(Xtest,ytest))

### Let's also examine boosting our models with AdaBoosting

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

for model in models:
    ada = AdaBoostClassifier(model,n_estimators=100,algorithm='SAMME.R',learning_rate=0.5,random_state=42)
    ada.fit(Xtrain,ytrain)
    print('model:',model)
    print('training score:',ada.score(Xtrain,ytrain))
    print('testing score:', ada.score(Xtest,ytest))

### And here we also try making a GradientBoosting model:

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

for depth in (2,4,6,8):
    for rate in (0.001,0.01,0.1):
        gbModel = GradientBoostingClassifier(max_depth=depth,learning_rate=rate,random_state=0)
        gbModel.fit(Xtrain,ytrain)
        print('depth:',depth)
        print('learning rate:',rate)
        print('training score:',gbModel.score(Xtrain,ytrain))
        print('testing score:',gbModel.score(Xtest,ytest))

### With a learning rate of 0.01 and a max_depth of 2, we have our best initial gradient boosting model, with a training score of 90.8% and a testing score of 78.1%.

In [ ]:
gbModelFinal = GradientBoostingClassifier(max_depth=2,learning_rate=0.01,random_state=0)
gbModelFinal.fit(Xtrain,ytrain)
print('training score:',gbModelFinal.score(Xtrain,ytrain))
print('testing score:',gbModelFinal.score(Xtest,ytest))

### DON'T FORGET TO TALK ABOUT WHETHER OR NOT BOOSTING AND BAGGING HELPED YOUR MODELS AND WHY

### Data reduction using PCA using SVD decomp

In [ ]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 50)
dataRed = pca.fit_transform(Xdata)
print('# of components:',10)
print(pca.explained_variance_ratio_)

In [ ]:
var = pca.explained_variance_ratio_
var.sum()
dataRed.shape

### By reducing the dataset from 270+ dimensions down to 50, we still preserve 97.6% of the variance and therefore explanatory power of our data for our models. Now we will try and run our models upon this new dataset.

In [ ]:
pcaTrain,pcaTest,ytrain,ytest=train_test_split(dataRed,ydatamulti,random_state=0)

model_names = ['logit','rForest','KSVM','LSVM','Tree','KNN']
for i in range(5):
    models[i].fit(pcaTrain,ytrain)
    print('model:',model_names[i])
    print('training score:',models[i].score(pcaTrain,ytrain))
    print('testing score:',models[i].score(pcaTest,ytest))

### As expected, we see a decrease in performance across all models using PCA transformed data. This is expected because, ideally, we would never want to use PCA, since we lose variance in our dataset during the reduction in dimensions (not to mention we lose the interpretability of the model). The only reason to use PCA transformation on our data is to reduce the processing load on our computers while losing the minimal amount of explanatory power of our models as possible. However, since we can run the original models just fine, we don't want to be using PCA here.

In [ ]:
models[0]

In [ ]:
#models = [logitFinal,rforestFinal,KSVMFinal,LSVMFinal,treeFinal,knnFinal]
